### ISSUE TITLE
Fine-tuning Model with Hard Negative Mining is Unexpectedly Slow

### ISSUE THREAD

**Issue by Iambestfeed** [UKPLab/sentence-transformers#3284]
#### **Description**  
I am experiencing unexpectedly slow training when fine-tuning a model using a **hard negative mining dataset**. My dataset consists of **70k samples**, formatted as:  
`query + pos + neg + neg_1 + neg_2 ... neg_63`

The training speed is significantly slower compared to using `MultipleNegativesRankingLoss` with a **batch size of 64** and **300k samples**.  

---

#### **Observed Training Speed**  
- **Current Training (Hard Negative Mining, 70k samples)**:  
  ```
  0%| | 33/37310 [00:45<13:57:29, 1.35s/it]
  ```
- **Baseline Training (MultipleNegativesRankingLoss, 300k samples)**:  
  ```
  50%|███████████████████▌ | 2732/5464 [54:42<53:38, 1.18s/it]
  ```  

🚨 **Issue:** Training with hard negative mining takes **~1.35s per iteration**, whereas the baseline method takes **~1.18s per iteration**—despite the dataset being much smaller.  

---

#### **Training Metrics (Hard Negative Mining, 70k samples)**  
```python
{'loss': 2.5917, 'grad_norm': 16.541519165039062, 'learning_rate': 4.288394532296972e-08, 'epoch': 0.0}
{'loss': 2.4103, 'grad_norm': 22.089021682739258, 'learning_rate': 9.112838381131065e-08, 'epoch': 0.0}
{'loss': 1.982, 'grad_norm': 18.620956420898438, 'learning_rate': 1.447333154650228e-07, 'epoch': 0.0}
{'loss': 2.4772, 'grad_norm': 21.30409049987793, 'learning_rate': 1.9833824711873497e-07, 'epoch': 0.0}
{'loss': 2.3306, 'grad_norm': 42.66603469848633, 'learning_rate': 2.519431787724471e-07, 'epoch': 0.0}
  0%|                                     | 51/37310 [01:07<13:28:11,  1.30s/it]
```

---

#### **Expected Behavior**  
Training should be more efficient, especially considering the dataset size is much smaller than the baseline. The performance drop is unexpected.  

#### **Steps to Reproduce**  
1. Prepare a dataset with **70k samples**, structured as: `query + pos + neg + neg_1 + neg_2 ...`  
2. Fine-tune a model using **hard negative mining**.  
3. Compare training speed to **MultipleNegativesRankingLoss** (batch size 64, 300k samples).  

---

#### **Environment Details**  
- **Batch Size**: 64  
- **Dataset**: 70k samples (with hard negatives)  

Would appreciate any insights on what could be causing this slowdown! 🚀

Training scripts:
```python
# Hyper Params
CONFIG = {
    "base_model_path": "vinai/phobert-base-v2",

    "training_data": "iambestfeed/vnexpress-data-similarity",
    "data_name": "filtered_mining_data",
    "deepspeed_config_path": "deepspeed.json",
    "keep_in_memory": True,

    "batch_size": 1,
    "num_negatives" : 32,
    "num_epochs":1,
    "learning_rate": 2e-5,

    "output_dir":f"{BASE_DIR}/model_checkpoints",
    "logging_dir":f"{BASE_DIR}/train_logs",
}

class KTupleLoss(nn.Module):
    def __init__(
        self, model: SentenceTransformer, scale: float = 20.0
    ) -> None:
        super().__init__()
        self.model = model
        self.scale = scale
        self.cross_entropy_loss = nn.CrossEntropyLoss()

    def forward(self, sentence_features: Iterable[dict[str, Tensor]], labels: Tensor) -> Tensor:
        embeddings = [self.model(sentence_feature)["sentence_embedding"] for sentence_feature in sentence_features]

        # Collect the anchor, positive, and negative embeddings
        anchor_embeddings = embeddings[0]  # [batch_size, embedding_dim]
        positive_embeddings = embeddings[1]  # [batch_size, embedding_dim]
        negative_embeddings = torch.stack(embeddings[2:], dim=1)  # [batch_size, num_negatives, embedding_dim]

        # Normalize them
        anchor_embeddings = torch.nn.functional.normalize(anchor_embeddings, p=2, dim=-1)
        positive_embeddings = torch.nn.functional.normalize(positive_embeddings, p=2, dim=-1)
        negative_embeddings = torch.nn.functional.normalize(negative_embeddings, p=2, dim=-1)

        # Compute the similarity scores, i.e. 1) pairwise cosine similarity between anchor and positive,
        # and 2) pairwise cosine similarity between anchor and negatives
        pos_similarity = (anchor_embeddings * positive_embeddings).sum(1, keepdim=True) # [batch_size, 1]
        anchor_embeddings_3d = anchor_embeddings.unsqueeze(1) # [batch_size, 1, embedding_dim]
        neg_similarity = torch.matmul(anchor_embeddings_3d, negative_embeddings.transpose(1, 2)).squeeze(1) # [batch_size, num_negatives]

        # Concatenate the positive and negative similarity scores so we have 1 + num_negatives similarity scores per anchor
        scores = torch.cat((pos_similarity, neg_similarity), dim=1) * self.scale # [batch_size, 1 + num_negatives]

        # Set the labels as 0, i.e. the positive sample is always the first one in the scores tensor
        labels = torch.zeros(scores.size(0), dtype=torch.long, device=scores.device)

        return self.cross_entropy_loss(scores, labels)

    def get_config_dict(self) -> dict[str, Any]:
        return {"scale": self.scale}


logging.basicConfig(format="%(asctime)s - %(message)s", datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO)  

logging.info(CONFIG)


def main():
    # Set the log level to INFO to get more information  

    logging.info("=== Starting training script ===")
    
    logging.info(CONFIG)
    

    # Load dataset
    logging.info("Loading dataset...")
    final_train_dataset = load_dataset(CONFIG['training_data'], CONFIG['data_name'])['train']
    logging.info(f"Dataset loaded: {final_train_dataset}")
    logging.info(f"First sample: {final_train_dataset[0]}")

    columns_to_keep = ["anchor", "positive", "negative"] + [f"negative_{i}" for i in range(1, CONFIG['num_negatives'] - 1)]
    final_train_dataset = final_train_dataset.select_columns(columns_to_keep)

    if TRIAL:
        logging.warning("Running in TRIAL mode! Using only 300 samples.")
        final_train_dataset = final_train_dataset.select(range(min(300, len(final_train_dataset))))

    # Dọn dẹp bộ nhớ
    gc.collect()
    torch.cuda.empty_cache()

    # Load model
    logging.info("Loading model for finetuning...")
    model = SentenceTransformer(CONFIG["base_model_path"], trust_remote_code=True)
    model.max_seq_length=256
    logging.info(f"Final Finetuning Model:\n{model}")

    # Định danh tên model trên Hugging Face Hub
    huggingface_model_name = f"iambestfeed/{CONFIG['base_model_path'].split('/')[1]}-finetuned-{CONFIG['data_name']}-lr{CONFIG['learning_rate']:.0e}-{CONFIG['num_epochs']}-epochs-bs-{CONFIG['batch_size']}"
    logging.info(f"Model name on Hugging Face Hub: {huggingface_model_name}")

    args = SentenceTransformerTrainingArguments(
        do_train=True,
        # Required parameter:
        output_dir=CONFIG["output_dir"],
        overwrite_output_dir = False,
        # Optional training parameters:
        num_train_epochs=CONFIG['num_epochs'],
        per_device_train_batch_size=CONFIG["batch_size"],
        learning_rate=CONFIG["learning_rate"],
        warmup_ratio=0.1,
        fp16=True,  # Set to False if you get an error that your GPU can't run on FP16
        batch_sampler=BatchSamplers.NO_DUPLICATES,
        # Optional tracking/debugging parameters:
        max_grad_norm = 1.0,
        #deepspeed = CONFIG["deepspeed_config_path"],
        disable_tqdm=False,
        logging_dir = CONFIG["output_dir"],
        save_strategy="epoch",  # Không lưu checkpoints
        save_total_limit=4,  # Chỉ giữ lại 1 model
        logging_steps=10,
        hub_model_id=huggingface_model_name,
        hub_strategy = 'every_save',
        report_to="tensorboard",
        push_to_hub = True,
        save_safetensors=False,
        torch_empty_cache_steps=None,
        load_best_model_at_end = False,
        # ignore_data_skip=True
    )

    train_loss = KTupleLoss(model)

    trainer = SentenceTransformerTrainer(
        model=model,
        args=args,
        train_dataset=final_train_dataset,
        loss=train_loss,
    )
    trainer.train()

    model.push_to_hub(huggingface_model_name, exist_ok = True)


if __name__ == "__main__":
    main()
```

**Comment by tomaarsen** [comment]
Hello!

Am I correct in understanding that the `MultipleNegativesRankingLoss` setting uses 300k query-answer tuples, i.e. not n-tuples with 64 negatives?

If so, then I'll try and elaborate:
### 300k pairs with `MultipleNegativesRankingLoss` (MNRL)
With a batch size of 64, every batch consists of 128 texts: 64 queries and 64 answers. We have to embed each of these 128 texts, and then we create a similarity matrix between the queries and the "candidates" (i.e. the texts from which the model has to identify the correct positive). In MNRL, this is all answers and negatives, a.k.a. the 64 answers in this case. Our matrix has shape (64, 64), and for each row $i$ (i.e. query $i$), candidate $i$ is the correct one. The Cross Entropy Loss then computes a loss value with this idea.

So: 300k pairs means having to compute 600.000 embeddings, and each batch involves creating a (64, 64) matrix (which is probably pretty quick, no big deal there).

### 70k with hard negatives
We still have a batch size of 64, but now every batch consists of 64 queries, 64 answers, and 64 * 64 = 4096 negatives - 4224 texts to embed in total. We again create a similarity matrix between the queries and the candidates, but now the candidates are not just the 64 answers, but *also* each of the 4096 negatives (a.k.a. 4160 total candidates). Yes, like before, that means that the candidates for query $i$ also involve the answer and negatives mined for query $j$, this is called in-batch negatives. So, our similarity matrix becomes (64, 4160). The model is still tasked with finding the correct answer (i.e. query $i$ has the correct answer at index $i$)  out of this huge matrix.

So: 70k n-tuples means having to compute 4.620.000 embeddings, and each batch involves creating a (64, 4160) matrix (which does sound pretty slow and memory-intensive).

### In short
So although we have less samples (70k vs 300k), we actually have to compute 7.7x more embeddings! Beyond that, it has more heavy operations like the similarity matrix creation, and the cross entropy loss computation is also probably slower.

You can go 1 of 2 routes:
1. Update the loss function to no longer use in-batch negatives. This means the similarity matrix becomes (64, 65) (just the positive and mined hard negatives for that query). This avoids the big matrix, but you still need to embed the 4.6m embeddings, so it'll still be slow. You also lose the "random negatives" which can help with performance.
2. Use `dataset = dataset.select_columns(["anchor", "positive", "negative", "negative_1", ..., "negative_5"])`, i.e. only use a handful of hard negatives. E.g. if you use 8 hard negatives instead of 64, then you'll be back down to 700.000 embeddings to compute (a bit like the 300k setting which needed 600k), and your similarity matrix will be (64, 640) which is also more manageable.

I would recommend option 2. I think it should bring the training time back down to be ~equivalent as the 300k setting.

- Tom Aarsen

**Comment by Iambestfeed** [comment]
> 2. Use `dataset = dataset.select_columns(["anchor", "positive", "negative", "negative_1", ..., "negative_5"])`, i.e. only use a handful of hard negatives. E.g. if you use 8 hard negatives instead of 64, then you'll be back down to 700.000 embeddings to compute (a bit like the 300k setting which needed 600k), and your similarity matrix will be (64, 640) which is also more manageable.
Sorry for not reading carefully. I looked it up and understood my misunderstanding.

Besides that, I have 2 issues that I hope you can help me with.

1. Does using 5 negatives seem bad? (are they too easy?)
2. I am training on TPU using model scripts https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2
```python
def train_function(index, args, queue):
    tokenizer = AutoTokenizer.from_pretrained(args.model)
    model = AutoModelForSentenceEmbedding(args.model, tokenizer)
    
  
    ### Train Loop
    device = xm.xla_device()
    model = model.to(device)

    # Instantiate optimizer
    optimizer = AdamW(params=model.parameters(), lr=2e-5, correct_bias=True)

    lr_scheduler = get_linear_schedule_with_warmup(
        optimizer=optimizer,
        num_warmup_steps=500,
        num_training_steps=args.steps,
    )
    
    # Now we train the model
    cross_entropy_loss = nn.CrossEntropyLoss()
    max_grad_norm = 1

    model.train()
   
    for global_step in tqdm.trange(args.steps, disable=not xm.is_master_ordinal()):
        #### Get the batch data
        batch = queue.get()
        #print(index, "batch {}x{}".format(len(batch), ",".join([str(len(b)) for b in batch])))
        

        if len(batch[0]) == 2: #(anchor, positive)
            text1 = tokenizer([b[0] for b in batch], return_tensors="pt", max_length=args.max_length, truncation=True, padding="max_length")
            text2 = tokenizer([b[1] for b in batch], return_tensors="pt", max_length=args.max_length, truncation=True, padding="max_length")

            ### Compute embeddings
            embeddings_a = model(**text1.to(device))
            embeddings_b = model(**text2.to(device))
            
            ### Gather all embedings 
            embeddings_a = torch_xla.core.functions.all_gather(embeddings_a)
            embeddings_b = torch_xla.core.functions.all_gather(embeddings_b)

            ### Compute similarity scores 512 x 512
            scores = torch.mm(embeddings_a, embeddings_b.transpose(0, 1)) * args.scale
        
            ### Compute cross-entropy loss
            labels = torch.tensor(range(len(scores)), dtype=torch.long, device=embeddings_a.device)  # Example a[i] should match with b[i]
            
            ## Symmetric loss as in CLIP
            loss = (cross_entropy_loss(scores, labels) + cross_entropy_loss(scores.transpose(0, 1), labels)) / 2

        else:   #(anchor, positive, negative)
            text1 = tokenizer([b[0] for b in batch], return_tensors="pt", max_length=args.max_length, truncation=True, padding="max_length")
            text2 = tokenizer([b[1] for b in batch], return_tensors="pt", max_length=args.max_length, truncation=True, padding="max_length")
            text3 = tokenizer([b[2] for b in batch], return_tensors="pt", max_length=args.max_length, truncation=True, padding="max_length")

            embeddings_a  = model(**text1.to(device))
            embeddings_b1 = model(**text2.to(device))
            embeddings_b2 = model(**text3.to(device))

            embeddings_a  = torch_xla.core.functions.all_gather(embeddings_a)
            embeddings_b1 = torch_xla.core.functions.all_gather(embeddings_b1)
            embeddings_b2 = torch_xla.core.functions.all_gather(embeddings_b2)

            embeddings_b = torch.cat([embeddings_b1, embeddings_b2])

            ### Compute similarity scores 512 x 1024
            scores = torch.mm(embeddings_a, embeddings_b.transpose(0, 1)) * args.scale
        
            ### Compute cross-entropy loss
            labels = torch.tensor(range(len(scores)), dtype=torch.long, device=embeddings_a.device)  # Example a[i] should match with b[i]
            
            ## One-way loss
            loss = cross_entropy_loss(scores, labels)

        
        # Backward pass
        optimizer.zero_grad()
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
        
        xm.optimizer_step(optimizer, barrier=True)
        lr_scheduler.step()


        #Save model
        if (global_step+1) % args.save_steps == 0:
            output_path = os.path.join(args.output, str(global_step+1))
            xm.master_print("save model: "+output_path)
            model.save_pretrained(output_path)
          
            
    output_path = os.path.join(args.output, "final")
    xm.master_print("save model final: "+ output_path)
    model.save_pretrained(output_path)
```

Do I need to customize the training function or can I apply MultipleNegativesRankingLoss directly?
And if so, I hope you can help me with this training script.

**Comment by tomaarsen** [comment]
No need to apologize! These loss functions can be very confusing, it's very understandable to get a bit mixed up.

1. No, 5 negatives is not bad. The most common settings are 0 or 1 negative per query, alongside the automatic in-batch negatives that you get. You can make them as hard or soft as you want.
2. That script is quite a manual implementation of the MNRL loss, and it only covers the (anchor, positive) and (anchor, positive, negative) setting, whereas you have 5 negatives. You should be able to use the MNRL loss to get the same behavior (except it'll work with your 5 negatives (or however many you want to use, it's up to you :) )). I would not recommend starting with the script used to train [all-MiniLM-L6-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2), you'll have to do a lot manually.

- Tom Aarsen

**Comment by Iambestfeed** [comment]
> 2. That script is quite a manual implementation of the MNRL loss, and it only covers the (anchor, positive) and (anchor, positive, negative) setting, whereas you have 5 negatives. You should be able to use the MNRL loss to get the same behavior (except it'll work with your 5 negatives (or however many you want to use, it's up to you :) )). I would not recommend starting with the script used to train [all-MiniLM-L6-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2), you'll have to do a lot manually.

Well, to be honest, since I use TRC (TPU Research Cloud), I need to use the code for TPU. Looks like I'll give it a try (although I'm sure it won't be easy :))) ).

But I looked at the training code of all-MiniLM-L6-v2 and I saw the file [data.config](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/blob/main/data_config.json). I wonder if there's a strategy for how you arrange their weights? I've experimented and found out that the order does matter a lot but I haven't seen any paper or blog that explains this so I'm a bit confused.

**Comment by tomaarsen** [comment]
I don't actually know how my predecessor came up with the data weights, I'm afraid. I'm also not very familiar with TPU units, but based on this blogpost (https://www.philschmid.de/getting-started-tpu-transformers) from Philipp Schmid, I think `transformers` and `accelerate` (which are used in Sentence Transformers) should take care of the TPU side of things for you. In short, I think the ["normal" training code](https://github.com/UKPLab/sentence-transformers/tree/master/examples/sentence_transformer/training) might just work out of the box? I can't be sure, though!

- Tom Aarsen

**Comment by Iambestfeed** [comment]
> I don't actually know how my predecessor came up with the data weights, I'm afraid. I'm also not very familiar with TPU units, but based on this blogpost (https://www.philschmid.de/getting-started-tpu-transformers) from Philipp Schmid, I think `transformers` and `accelerate` (which are used in Sentence Transformers) should take care of the TPU side of things for you. In short, I think the ["normal" training code](https://github.com/UKPLab/sentence-transformers/tree/master/examples/sentence_transformer/training) might just work out of the box? I can't be sure, though!


Hmm, thanks for your support. I think the best way is to implement everything myself. 
Thank you very much.
